### PR TITLE
Added sort by for groups by creation date and member groups.

### PR DESCRIPTION
**Pull request by skv93-coder**
Date: 25-12-23

Developer Name: @skv93-coder 

----

## Issue Ticket Number

-   https://github.com/Real-Dev-Squad/website-backend/issues/1707

## Description

This PR adds sort by 'Creation date' and 'member count' for groups

**Documentation Updated?**

-   [ ] Yes
-   [x] No

_If your feature adds a new API, then documentation related to the feature like API-Contracts, and Data-models  must be updated._

**Breaking Changes**

-   [ ] Yes
-   [x] No

_If your feature introduces breaking changes or if something is missing, please mention the related issue tickets._

**Development Tested?**

-   [x] Yes
-   [ ] No

_Confirm whether the changes have been tested locally during development._

**Tested in Staging?**

-   [ ] Yes
-   [x] No

_Indicate whether the changes have been tested in the staging environment._

**Under Feature Flag**

-   [x] Yes
-   [ ] No

_Specify if the changes are currently under a feature flag._

**Database Changes**

-   [ ] Yes
-   [x] No

_Indicate whether the changes include modifications to the database._

### Screenshots

[screen-capture (5).webm](https://github.com/Real-Dev-Squad/website-dashboard/assets/68867418/709e6d62-b6c1-48af-9bd0-a295cd89fd81)

## Test Coverage

image
image

### CODE REVIEW

File: __tests__/groups/group.test.js [hunk 1]
```diff
@@ -349,4 +349,55 @@ describe('Discord Groups Page', () => {
     );
     expect(groupName).toMatch('DSA');
   });
+  test('On click on "Popular with dev" will result group with most member at the top', async () => {
+    await page.goto('http://localhost:8000/groups?dev=true');
```
Comment (ajoykumardas12) [inline __tests__/groups/group.test.js:353]: Should use BASE_URL variable.

File: __tests__/groups/group.test.js [hunk 2]
```diff
@@ -349,4 +349,55 @@ describe('Discord Groups Page', () => {
...
+    const groupsAfterSort = await page.$$eval('.group-name', (elements) => {
+      return elements.map((element) =>
+        element.getAttribute('data-member-count'),
+      );
+    });
+    const manualSortedGroup = groupsBeforeSort.sort((a, b) => b - a);
+    expect(groupsAfterSort).toEqual(manualSortedGroup);
+  });
+  test('On click on "Recently created" will result in latest created group at the top', async () => {
+    await page.goto('http://localhost:8000/groups?dev=true');
```
Comment (ajoykumardas12) [inline __tests__/groups/group.test.js:377]: same as above.

File: groups/constants.js [hunk 3]
```diff
@@ -1,5 +1,14 @@
 const NO_SPACES_ALLOWED = 'Roles cannot have spaces';
 const CANNOT_CONTAIN_GROUP = "Roles cannot contain 'group'.";
 const DEV_FEATURE_FLAG = 'dev';
+const SortByFields = [
+  { fieldName: 'memberCount', id: '1' },
+  { fieldName: 'date._seconds', id: '2' },
```
Comment (ajoykumardas12) [inline groups/constants.js:6]: NIT: having id declared first is more readable.

File: groups/index.html [hunk 4]
```diff
@@ -36,6 +36,13 @@ <h3 class="not-verified-tag hidden">
           <ul class="groups-list"></ul>
         </aside>
         <main>
+          <div class="dropdown hidden" id="dropdown_container">
+            <button class="task_filter-button" id="sortby_text">Sort by</button>
+            <div id="dropdown_main" class="dropdown-content">
+              <a data-list="1">Popular with dev</a>
```
Comment (ajoykumardas12) [inline groups/index.html:42]: There can be users who aren't developers, so not sure about this. Please confirm with Ankush/Tejas if hasn't been.

File: groups/script.js [hunk 5]
```diff
@@ -22,6 +23,19 @@ const userIsNotVerifiedText = document.querySelector('.not-verified-tag');
...
+//Dropdown
+const dropdown_main = document.getElementById('dropdown_main');
+const dropdown_txt = document.getElementById('sortby_text');
+function toggleDropDown() {
+  dropdown_main.classList.toggle('show_filter');
+}
+dropdown_txt.addEventListener('click', toggleDropDown);
+if (isDev) {
+  dropdownContainer.classList.remove('hidden');
+  dropdown_main.addEventListener('click', onDropdownClick);
```
Comment (ajoykumardas12) [inline groups/script.js:37]: Why not using camelCase for some variables?

File: groups/script.js [hunk 6]
```diff
@@ -76,46 +90,85 @@ const memberAddRoleBody = {
  *
  * FOR RENDERING GROUP ROLES IN 'MANAGE ROLES' TAB
  */
-const groupsData = await getDiscordGroups();
+let groupsData = await getDiscordGroups();
```
Comment (ajoykumardas12) [inline groups/script.js:93]: Why using `let` ? Don't see the variable changing.

File: groups/script.js [hunk 7]
```diff
@@ -76,46 +90,85 @@ const memberAddRoleBody = {
...
+      secondObj,
+      fieldToSortBy.fieldName,
+    );
+    if (a > b) {
+      return -1;
+    } else if (b < a) {
+      return 1;
+    }
+    return 0;
+  });
```
Comment (ajoykumardas12) [inline groups/script.js:167]: Would something like this work here?
```groupsData.sort((firstObj, secondObj) => firstObj[fieldToSortBy.fieldName] - secondObj[fieldToSortBy.fieldName])```

The code can be more concise this way.

File: groups/style.css [hunk 8]
```diff
@@ -284,6 +284,72 @@ NOT VERIFIED TEXT ABOVE
   text-align: center;
 }
 
+.dropbtn {
```
Comment (ajoykumardas12) [inline groups/style.css:287]: `dropdown-btn` would be easier to read.

File: groups/style.css [hunk 9]
```diff
@@ -284,6 +284,72 @@ NOT VERIFIED TEXT ABOVE
...
+}
+
+.dropdown-content {
+  display: none;
+  position: absolute;
+  background-color: #f6f6f6;
+  min-width: 14rem;
+  overflow: auto;
+  border: 1px solid #ddd;
+  z-index: 1;
```
Comment (ajoykumardas12) [inline groups/style.css:306]: Use variable for colors.

File: groups/style.css [hunk 10]
```diff
@@ -284,6 +284,72 @@ NOT VERIFIED TEXT ABOVE
...
+  color: #000;
+  font-weight: 700;
+  cursor: pointer;
+  border: none;
+  outline: none;
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+  padding-left: 0.3rem;
+  border: 1px solid #d0d7de;
```
Comment (ajoykumardas12) [inline groups/style.css:342]: border is set two times.

File: groups/script.js (hunk 7 above)
Comment (ajoykumardas12) [reply groups/script.js:167]: Okay, this doesn't work because of the field might not be present case.

File: groups/style.css (hunk 10 above)
Comment (skv93-coder) [reply groups/style.css]: Thanks, removed redundant prop

File: groups/style.css (hunk 9 above)
Comment (skv93-coder) [reply groups/style.css]: Done

File: groups/style.css (hunk 8 above)
Comment (skv93-coder) [reply groups/style.css]: redundant class, so removed

File: groups/script.js (hunk 6 above)
Comment (skv93-coder) [reply groups/script.js]: Done.

File: groups/script.js (hunk 5 above)
Comment (skv93-coder) [reply groups/script.js]: Done

File: groups/index.html (hunk 4 above)
Comment (skv93-coder) [reply groups/index.html:42]: This UI is approved [design doc](https://docs.google.com/document/d/1SqHcsh-KCBewySP1rCx8k2AJqfdsQJMaXjAxMkHUrKY/edit#heading=h.cxpij3s7de98)

File: groups/constants.js (hunk 3 above)
Comment (skv93-coder) [reply groups/constants.js]: Sure, done

File: groups/index.html (hunk 4 above)
Comment (skv93-coder) [reply groups/index.html]: Changes the text after suggestion from tejas.

**Review by ajoykumardas12 — APPROVED** [review]

**Pull request merged**